docs/motivation.md

```diff
@@ -0,0 +1,535 @@
+```verona
+searchGraph[V, N, E](n : N, v : V) : Option[N]
+    where ValueGraph[V, N, E] & (V <: Equal[V]) & (N <: Comparable[N])
+{
+    let visited = Set[N]::create();
+    let pending = Queue[N]::create({n});
+
+    while (!pending.empty()) {
+        let next = pending.dequeue();
+
+        if (!visited.contains(next)) {
+            visited.add(next);
+            if (next.val().equals(v)) {
+                Some(next)
+            }
+            else {
+                for (edge : next.edges()) {
+                    pending.enqueue(edge.dst());
+                }
+            }
+        }
+    }
+
+    None
+}
+```
+
+#### Method-level `where` clauses
+Adding type constraints to methods, we can condition the existance of a method
+depending on the specific type parameters. Going back to the red-black tree
+example above, we could add a `print` method that will exist only when the
+type parameter to `RBTree` implements the `Printable` trait:
+```verona
+class RBTree[T] where (T <: Comparable[T]) {
+    ...
+    print(self: Self) : Unit where (T <: Printable) {
+        ...
+    }
+}
+```
+
+##### `where` clauses and capabilities
+Another use for method-level where clauses is to restrict what methods can be
+called depending the capabilities of type parameters.
+```verona
+class Ref[T] {
+    var val : T
+    get(self: Self) : T where T <: mut | imm // i.e. not iso
+    {
+        val
+    }
+    set(self: Self, v : T) : T { val = v }
+}
+
+class Array[T] {
+    // uses Ref[T] internally
+    ...
+    get(self: Self, idx : U32) : T where T <: mut | imm // i.e. not iso
+    {
+        ...
+    }
+    set(self: Self, idx : U32, v : T) : T { ... }
+}
+
+class HashMap[K, T] {
+    ... // built on top of Array[T]
+    get(self : Self, k : K) : T where T <: mut | imm // i.e. not iso
+    {
+        ...
+    }
+    set(self : Self, k : K, v : T) : T { ... }
+}
+```
+In a system without method-level `where` clauses, we would need to bifurcate each
+class on whether the class should be able to handle iso values or not, e.g.
+having to define both `IsoRef[T]` and `Ref[T]`. Furthermore, `where` clauses on
+methods are very intentional, and allows us to give better error messages. E.g.
+"`Ref[iso & A]::get()` not available since `iso & A </: imm | mut`" will
+hopefully help the programmer more than "`Ref[iso & A]` is not a valid
+instantiation of `Ref[T]`".
+
+##### Default methods and type predicates
+Together with type predicates we might be able to split part of `ToString` (see
+[Trait Inheritance](#trait-inheritance)) into a `Printable` trait. `Printable`
+includes a default implementation in the case where the concrete type also
+implements the `ToString` trait:
+```verona
+type ToString = {
+    toString(s: Self) : String
+}
+
+type Printable = {
+    print(s: Self) : Unit // this will have to be defined if Self </: ToString
+    print(s: Self) : Unit where (Self <: ToString) {
+        sys.io.out(s.toString())
+    }
+}
```

Comment:
Is there always a most specific constraint? There is a parallel to pattern matching where you can have a matching order that does not depend on the order that cases are written, but on their specificity, and then you need to look out for patterns that cannot be ordered, e.g., the patterns for tuples of lists `(nil, l2)` and `(l1, nil)`. I wonder if a similar problem can appear here.

Reply:
don't know, we should discuss later